source/logic.rst

```diff
@@ -0,0 +1,1223 @@
+since then you can use the triangle inequality.
+Notice that we marked all the variables ``s``, ``t``, ``a``, and ``b``
+implicit because they can be inferred from the hypotheses.
+
+Proving the same theorem with multiplication in place
+of addition is tricky.
+We will get there by proving some auxiliary statements first.
+See if you can also finish off the next proof,
+which shows that if ``s`` converges to ``a``,
+then ``λ n, c * s n`` converges to ``c * a``.
+It is helpful to split into cases depending on whether ``c``
+is equal to zero or not.
+Since we have not shown you how to do that yet,
+we have taken care of the zero case first,
+and we have left you to prove the result with
+the extra assumption that ``c`` is nonzero.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    theorem converges_to_const (a : ℝ) : converges_to (λ x : ℕ, a) a :=
+    sorry
+
+    variables {s : ℕ → ℝ} {a : ℝ}
+
+    -- BEGIN
+    theorem converges_to_mul_const
+      {c : ℝ} (cs : converges_to s a) :
+    converges_to (λ n, c * s n) (c * a) :=
+    begin
+      by_cases h : c = 0,
+      { convert converges_to_const 0,
+        { ext, rw [h, zero_mul] },
+        rw [h, zero_mul] },
+      have acpos : 0 < abs c,
+        from abs_pos_of_ne_zero h,
+      sorry
+    end
+    -- END
+
+The next theorem is also independently interesting:
+it shows that a convergent sequence is eventually bounded
+in absolute value.
+We have started you off; see if you can finish it.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    variables {s : ℕ → ℝ} {a : ℝ}
+
+    -- BEGIN
+    theorem exists_abs_le_of_converges_to (cs : converges_to s a) :
+      ∃ N b, ∀ n, N ≤ n → abs (s n) < b :=
+    begin
+      cases cs 1 zero_lt_one with N h,
+      use [N, abs a + 1],
+      sorry
+    end
+    -- END
+
+The next lemma is auxiliary: we prove that if
+``s`` converges to ``a`` and ``t`` converges to ``0``,
+then ``λ n, s n * t n`` converges to ``0``.
+To do so, we use the previous theorem to find a ``B``
+on ``s`` beyond some point ``N₀``.
+See if you can understand the strategy we have outlined
+and finish the proof.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    variables {s t : ℕ → ℝ} {a : ℝ}
+
+    theorem exists_abs_le_of_converges_to (cs : converges_to s a) :
+      ∃ N b, ∀ n, N ≤ n → abs (s n) < b :=
+    sorry
+
+    -- BEGIN
+    lemma aux (cs : converges_to s a) (ct : converges_to t 0) :
+      converges_to (λ n, s n * t n) 0 :=
+    begin
+      intros ε εpos, dsimp,
+      rcases exists_abs_le_of_converges_to cs with ⟨N₀, B, h₀⟩,
+      have Bpos : 0 < B,
+        from lt_of_le_of_lt (abs_nonneg _) (h₀ N₀ (le_refl _)),
```

Comment:
Maybe it's time to explain `linarith`.

Reply:
We use `linarith` in the previous chapter and also this one. But here is is not a big win: the line becomes `{ linarith [abs_nonneg (s N₀), h₀ N₀ (le_refl _)] }`.